### ISSUE TITLE
Make *Async methods available for .NET 4.0 targeted projects.

### ISSUE THREAD

**Issue by Lumirris** [App-vNext/Polly#103]
Given that .NET 4.0 targeted projects can utilize Tasks and the await/async keywords through [Microsoft.Bcl.Async](https://www.nuget.org/packages/Microsoft.Bcl.Async/), it would be nice to make the *Async methods available for .NET 4.0 projects that use the Polly Nuget package.

**Comment by reisenberger** [comment]
@Lumirris Please feel free to take yourself a fork and try this out.  The availability / inclusion of the async overloads is controlled by a compilation constant `SUPPORTS_ASYNC`.  In the .NET4.0 proj, include [Microsoft.Bcl.Async](https://www.nuget.org/packages/Microsoft.Bcl.Async/), add `SUPPORTS_ASYNC` as a conditional compilation constant, and you should be good to go.

**Comment by reisenberger** [comment]
Hi @Lumirris Having consulted internally at App-vNext, current view is not to take extra dependencies in the core Polly project, as we're targeting .NET Core compatibility and reducing the overall number of separate projects / dependencies.  Will however leave the issue open to see if there is greater demand.

That said, it should be fairly easy to maintain your own fork for this and rebase it when you want to pick up new features.  

Happy for any feedback on how Polly async plays with [Microsoft.Bcl.Async](https://www.nuget.org/packages/Microsoft.Bcl.Async/) (shouldn't be any problems), from anything you try ...

**Comment by Lumirris** [comment]
I wonder if a simple solution would be to create another Nuget package like `Polly.Net40Async` that would leverage existing *Async code and use another project file like `src/Polly.Net45/Polly.Net40Async.csproj` which included 

`<DefineConstants>TRACE;DEBUG;`**SUPPORTS_ASYNC;**`SUPPORTS_READONLY_COLLECTION</DefineConstants>`

I haven't done something like this before, but if you think the concept is sound, I'll work on it. Does the idea of a separate Nuget package sounds good?

**Comment by reisenberger** [comment]
Hi @Lumirris   To make code suitable for a .NET4.0Async nuget package, I would follow the principles: 

+ only include a .NET4.0 proj (not 4.5, 3.5, PCL), so there isn't any confusion that this package provides anything other than specifically the .NET4.0Async version
+ only include one .NET4.0 proj in the package (wasn't sure if you were suggesting `src/Polly.Net45/Polly.Net40Async.csproj` (45 a typo?) as additional to the existing .NET40 package or replacement; suggest replacement).

Try the following steps:

+ take your own fork of Polly
+ remove all projects except for the .NET40 project, `Polly.Shared` and `Polly.SharedSpecs`
+ rename the .NET40 project in the solution as `Polly.Net40Async.csproj` as you suggest (makes it v clear)
+ reference [Microsoft.Bcl.Async](https://www.nuget.org/packages/Microsoft.Bcl.Async/) from that csproj
+ add `SUPPORTS_ASYNC` as a conditional compilation constant to that proj

Let us know how you get on.

**Comment by reisenberger** [comment]
Hey @Lumirris We talked further at AppvNext, I believe from @joelhulen we'd be happy to host and maintain (ie keep udpated) a Nuget package like `Polly.Net40Async`, as you suggest - assuming we can get all components to play.

@Lumirris Did you have any luck trying out a prototype, per previous notes?  (if you have an immediate need for this, you'd be well placed to test :smile: ).  Do you need any further assistance?  

@joelhulen I see this clearest as a separate package, not part of the main Polly package, to reduce the dependencies of the main Polly package.

Thanks!

**Comment by Lumirris** [comment]
@reisenberger @joelhulen I'll get working on this tonight and share a link to my fork before doing a pull request to fix up anything you spot that needs work.

**Comment by Lumirris** [comment]
@reisenberger @joelhulen I forgot about the fact that [Microsoft.Bcl.Async](https://www.nuget.org/packages/Microsoft.Bcl.Async/) gives you some of the async features through `TaskEx` vs. `Task`. The consequences of this are that the following lines of code in Polly.Shared would need to be changed just for this Nuget package, which would put it at odds with the other projects using these files:

 -SystemClock.cs ([here](https://github.com/App-vNext/Polly/blob/master/src/Polly.Shared/Utilities/SystemClock.cs#L33) and [here](https://github.com/App-vNext/Polly/blob/master/src/Polly.Shared/Utilities/SystemClock.cs#L53))
 -[RetryPolicyStateAsync.cs](https://github.com/App-vNext/Polly/blob/master/src/Polly.Shared/Retry/RetryPolicyStateAsync.cs#L11)
 -[RetryPolicyStateWithCountAsync.cs](https://github.com/App-vNext/Polly/blob/master/src/Polly.Shared/Retry/RetryPolicyStateWithCountAsync.cs#L13)

My thought is to replace calls to these async methods with calls to wrapping methods placed in a centralized location where the compiler constant switches are done, and utilize a new compiler constant, `SUPPORTS_ASYNC_40` to distinguish between .NET 4.5 async support and .NET 4.0 async support through the [Microsoft.Bcl.Async](https://www.nuget.org/packages/Microsoft.Bcl.Async/) package and `TaskEx`.

I'll post a link to my fork before making a PR for your review - please yank on the emergency brake if this seems like a bad path to take.

**Comment by reisenberger** [comment]
Hey @Lumirris.  Thanks for all your work on this!  

With you about `Task` vs `TaskEx` in Microsoft.Bcl.Async.  `SUPPORTS_ASYNC_40` switch sounds good.  

Am assuming (shout if I misunderstood) that `SUPPORTS_ASYNC_40` should _supplement_ (rather than replace) `SUPPORTS_ASYNC` as a compilation constant in the new package, so the `Polly.Net40Async.csproj` would set both  `SUPPORTS_ASYNC` and `SUPPORTS_ASYNC_40`.  This will certainly reduce modifications elsewhere.

Like the idea of centralising the `Task` vs `TaskEx` switches - just a slight tweak to suggest.  Rather than wrapping methods, can we follow the pattern that's already in SystemClock.cs?  IE Define a `public static Action<...>` and vary the definition of that.  So we might have a new class `TaskHelpers` in Utilities folder, the whole class definition wrapped in `#if SUPPORTS_ASYNC / #endif`, and then definitions like:

```csharp
        /// <summary>
        /// Allows the setting of a custom async Sleep implementation for testing.
        /// By default this will be a call to <see cref="M:Task.Delay"/>
        /// </summary>
#if !SUPPORT_ASYNC_40
        public static Func<TimeSpan, CancellationToken, Task> SleepAsync = Task.Delay;
#else
        public static Func<TimeSpan, CancellationToken, Task> SleepAsync = TaskEx.Delay;
#endif
```

@Lumirris What do you think?  Does this all sound good?

**Comment by joelhulen** [comment]
I really do like the pattern @reisenberger proposes here, which matches the existing SystemClock solution. This will cause the least impact on other areas of the solution. Either way, I really appreciate your contribution!

**Comment by reisenberger** [comment]
@Lumirris @joelhulen I put some investigation into the unit-testing for an async 4.0 package.

It turns out .NET4.0 unit tests were removed back in October 2015 because the latest XUnit 2.0+, which Polly uses, supports only 4.5+.  In other words, for the validity of the .NET4.0 and 3.5 packages we currently offer, we assume backwards .NET compatibility: that the supported .NET3.5 / .NET4.0 behaviours behave the same as the .NET4.5 and PCL tests verify.

Following the same logic, for the new .NET40async package, we would be relying additionally on the [Microsoft.Bcl.Async](https://www.nuget.org/packages/Microsoft.Bcl.Async/) `TaskEx` functionality behaving as the later `Task` functionality does.  Given the Microsoft.Bcl.Async package was the community technology preview for the relevant .Net4.5 elements anyway - and @Lumirris has identified the only switches will be around `Task.FromResult` and `Task.Delay` - I think this is acceptable.  Just @joelhulen sharing.

Alternatives could be ( ... but I propose we don't adopt either; they are not particularly palatable from the p-o-view of long-term maintenance ...) : 

+ Downgrade both the master Polly package and the .NET40 async package to the earlier XUnit 1.9.2, and remove the XUnit 2.0+ features we use.  Not proposing taking / holding the overall package back like this.
[or]
+ Maintain divergent sets of tests (on different versions of xunit) between the two different packages.  Maintenance headache.

If anyone has a brighter idea, please shout. :smile:

**Comment by Lumirris** [comment]
@reisenberger I'm in agreement with all of your recent suggestions pertaining to dealing with `Task` vs. `TaskEx`. I had started to go down that road already, and even experimented with using the compiler constant switch around the new `using static` syntax with `Task` and `TaskEx`, since the methods are named the same. Somehow that seems like an abuse, though, and brittle because it only works when the methods are named the same:
```c#
#if SUPPORTS_ASYNC
using System.Threading.Tasks;
#if SUPPORTS_ASYNC_NET40
using static System.Threading.Tasks.TaskEx;
#else
using static System.Threading.Tasks.Task;
#endif
#endif

namespace Polly.Utilities
{
    /// <summary>
    ///     Time-related delegates used to abstract away the async implementations
    /// </summary>
    public static class SystemClock
    {
#if SUPPORTS_ASYNC
        /// <summary>
        ///     Allows the setting of a custom async Sleep implementation for testing.
        ///     By default this will be a call to <see cref="M:Task.Delay" /> 
        ///     or <see cref="M:TaskEx.Delay" /> depending on the targeted framework.
        /// </summary>
        public static Func<TimeSpan, CancellationToken, Task> SleepAsync = Delay;
#endif
   }
}
```

**Comment by reisenberger** [comment]
@Lumirris  Huge thanks for offering to take on this Microsoft.Bcl.Async work!  How are you progressing with this?  Any blocks?  Do you need any assistance?  

@KBRETON1 Are you (having expressed an interest under #90) available to assist with this? (if we need any assistance).

@Lumirris Re my reference to this issue from #107: I have constructed the #108 fix for #107 so that users of Polly with .Net40 and Microsoft.Bcl.Async in VS2010 would be able to specify the new async onRetry delegates cleanly (by either of the two methods proposed in my comment https://github.com/App-vNext/Polly/issues/107#issuecomment-218895978), despite the VS2010 limitation.  So, my comments in #107 are not an issue for progressing this. 

Many thanks!

**Comment by KBRETON1** [comment]
@Lumirris is there a hosted branch that I could clone to get up to speed on this thread?  I am new to this project and willing to help.   I am in general agreement with the topics above.

**Comment by Lumirris** [comment]
@KBRETON1 I'm finishing up my work on this today; I'll push it up to my own fork, and have you and others review it before I do a pull request. Also, I see that my forked 'master' branch is now several commits behind the main 'master' branch. I'm new to open source contribution, and I'm not sure how I should proceed from here - I can't 'rebase' my forked master, since it too is public. Do I simply submit my pull request and @reisenberger would handle merge conflicts?

**Comment by joelhulen** [comment]
@Lumirris Thanks for your contribution! As far as handling merge conflicts and rebasing, that responsibility usually falls on the contributor submitting a pull request. The steps are simple and outlined (on our wiki page)[https://github.com/App-vNext/Polly/wiki/Git-Workflow#handling-updates-from-upstreammaster]. If you get stuck, simply reply to this thread and we'll do our best to help you sort it out.

**Comment by reisenberger** [comment]
@Lumirris Thanks for all your work on this, and @KBRETON1, thanks for your support/involvement!  @KBRETON1 Please (as another user wanting the .NET40 support) do feel free to review alongside me, at review stage.

@Lumirris For now, as you suggest, carry on pushing your changes to your master, to complete the work.  Will be at least clear to review there, before we engage in rebasing.    

@joelhulen The final requirement will be to publish a separate `Polly.Net40Async` nuget package.  That `Polly.Net40Async` nuget package will take the Microsoft.BCL.Async dependency, but, as we know, the main nuget Polly package _mustn't_.  Can you from your nuget script/process publish both packages like that out of the one GitHub repo?  Only asking as, if we have instead to split into two separate github repo sources (one without .NET40, one with), that may change the whole merge/rebase requirement anyway.

**Comment by joelhulen** [comment]
@reisenberger Yes, we can. You'll see within the `build.cake` file where the `__CreateNugetPackage` and `__CreateSignedNugetPackage` tasks are defined. We would need to define a new NuGet package task for `.Net40Async`, ensuring the appropriate project/assemblies are referenced.

I agree that it would be best to try and keep everything within this repo. Worst case, we create a new branch, but hopefully we won't have to go that route.

**Comment by reisenberger** [comment]
@joelhulen D'oh.  Don't know why I didn't check the `build.cake` for that.  Cool!  Imagine yes we can (and should) keep it all in the same repo then.

**Comment by Lumirris** [comment]
@reisenberger @joelhulen @KBRETON1 I've pushed the changes to my feature branch on my own fork. I'm not sure if you'll ultimately want the PR to come from my master branch or this feature branch, but I wanted to have the changes reviewed before the PR is made.

[Comparing App-vNext Polly (master) to Lumirris Polly (async-dot-net-40)](https://github.com/App-vNext/Polly/compare/master...Lumirris:async-dot-net-40)

**Comment by reisenberger** [comment]
@Lumirris If you have managed to move it off your master and into the feature branch (as looks), even better.  Great: will review!  Many thanks for this!

**Comment by Lumirris** [comment]
@reisenberger It turned into an even smaller change than before rebasing my master to incorporate recent PRs. There was only the [SystemClock](Polly/src/Polly.Shared/Utilities/SystemClock.cs) that needed any change. I noticed that the Visual Studio version changed slightly in the solution file (I think I've applied the most recent VS2015 update), which could be deleted if needed. Also, I'm having a hard time getting unit tests to be discovered, so I couldn't run any of the xUnit tests, but I'm guessing the build server will.

**Comment by reisenberger** [comment]
@Lumirris Re the tests, a version of XUnit was adopted a while back that no longer supports .NET40, so makes sense if no tests could be run in the version of the solution you've just pushed.  Ref my [earlier comment](https://github.com/App-vNext/Polly/issues/103#issuecomment-216357299): we'll rely on the .NET4.5 tests to assure the behaviour of the .NET4.0+Bcl.Async solution.  Given the tiny differences, this seems fine.

Great work on the rebase.

 >only the SystemClock that needed any change

Yeah, I squashed some `Task.FromResult(...)` stuff out when fixing the async onRetry delegs.

**Comment by reisenberger** [comment]
@Lumirris REVIEW: Given the way we have now decided to master the nuget packages from one repo (rather than a separate repo for this), the advice in my [25 April comment](https://github.com/App-vNext/Polly/issues/103#issuecomment-214204745) to remove all other csprojs (.NET35, .NET45, PCL, .NET45specs, PCLspecs) was as it turns out unhelpful (wrong in retrospect!).  Can you reinstate all the deletions of those from Polly.sln?  (Copy-paste from a good master or from the diff?)  Then I will pull down and build/test locally.

Shouldn't be further probs - the eventual changes are so small!

**Comment by Lumirris** [comment]
OK - the solution file has been updated in this [latest commit](https://github.com/Lumirris/Polly/tree/async-dot-net-40). Looking forward to hearing the results of local testing and what you'd like from me after that with regards to a PR.

**Comment by reisenberger** [comment]
@Lumirris In principle this is looking good!  Aiming for some brief real-world testing of the NET40-async package tomorrow / day after(?), then we'll sort out a PR and work on the `build.cake` nuget packaging our end.

@Lumirris Are there any unused references among the `Polly.Net40Async` project?  Could anything unused be removed?  Guessing we can't be using anything from `System.Net`, `System.IO` ... any others that can go? (possibly just adding Bcl async via nuget pulled these refs in...)

![pollynet40async-references](https://cloud.githubusercontent.com/assets/9608723/15591618/fc1e5b66-2395-11e6-920c-d8c431fb355a.PNG)

Many thanks for all your work on this!

**Comment by reisenberger** [comment]
@Lumirris. We need to adjust the documentation and start setting up the cake build to generate the new nuget packages for this.  We may need to experiment with the cake/nuget packaging a bit, so I'm going to pull (PR) your fork across to a fork on my repo reisenberger/Polly first, then start fixing things up there (cuts down toing-froing noise of requests to you).  I'll re-alert you when we have something ready to PR to app-vNext/master!

Thanks for all the initial work on this!  Hopefully get this out soon.

**Comment by reisenberger** [comment]
@Lumirris @joelhulen Please find a version with some doco added, nearly ready for PR, [here](https://github.com/App-vNext/Polly/compare/master...reisenberger:Net40Asyncv2?expand=1)

@Lumirris @KBRETON1 @joelhulen Anything need correcting?

@joelhulen `build.cake` and the nuget packaging next.

**Comment by reisenberger** [comment]
@joelhulen Over to you to fix up the nuget packaging side in `build.cake`.  

- I am fully signed off on the content.  
- .NET4.0 async tests fine with both manual tests and the most complicated example in Polly-samples.  

Joel: Please feel free to make a PR directly on to app-vNext/master [from here](https://github.com/App-vNext/Polly/compare/master...reisenberger:Net40Asyncv2?expand=1) when ready.

I assumed (already listed in the doco) `Polly.Net40Async` and `Polly.Net40Async-signed`.as the names for the new nuget packages.

@Lumirris Thanks for your contribution!  Pushed this through quickly today as am travelling and largely away from keyboard next 10 days.  Thanks!

**Comment by Lumirris** [comment]
@reisenberger Regarding the unused references, I got a chance to review this - I had used Resharper's 'Optimize References' on the project to detect what wasn't being used, but as you surmised, pulling in the Microsoft.Bcl.Async package brings a dependency on Microsoft.Bcl, which in turn brings in System.IO. 

However, it doesn't explain System.Net - When I use R# 'Remove Unused References', it didn't suggest removing that one, but yet it doesn't show any code dependent on it either. I decided not to remove it, believing that was the more conservative approach. Every time I try to get cute with stuff like that, I get burned.

I also don't know where the 'System.Core' reference came from, and why only this project needs it, but there was lots of code the referenced it, all where LINQ statements were used.

Anyway, thank you for taking over the PR from here. I'm looking forward to downloading the new package!

**Comment by KBRETON1** [comment]
@Lumirris @reisenberger It looks good to me.  I have used Microsoft.Bcl.Async in other projects.  The System.IO was added by default.  I was able to remove it and didn't have an issue at compile time.   I'll be running this new code on an XP box with .NET 4.0 in the next several days.

**Comment by KBRETON1** [comment]
@Lumirris @reisenberger, I forgot to add that the missing nuget.config file caused the packages to get downloaded to %appdata%\nuget folder instead of the solution\packages folder.  This caused my build to fail as the hint path referenced ..\packages folder.    Is there something else I should have configured to prevent this or should the nuget.config file be added?

**Comment by reisenberger** [comment]
@KBRETON1 Thanks for highlighting that potential issue with the Nuget.config.  What version of Visual Studio are you using?  I didn't have that issue (VS2015).  Nevertheless, I will look at the missing nuget.config in due course.  Thanks!

**Comment by KBRETON1** [comment]
@reisenberger, I am using VS2015.

**Comment by Lumirris** [comment]
@kbreton1 re: references, do you have any inside into why System.Net is there, yet is unused and why Sydtem.Core is there (and used extensively) but is not in the other Polly projects?

**Comment by reisenberger** [comment]
Thanks @KBRETON1.  On reflection, has to be VS2015, given the Shared projects in the Polly solution.

Re:

>do you have any inside into why System.Net is there?

Sure I'm stating the obvious here, but the classic use cases for async are to not block a thread while the system waits for a response from an IO device (methods like `System.Io.StreamReader.ReadToEndAsync()`), or while waiting for a network response (methods like `System.Net.Http.HttpClient.GetStringAsync()`).  It thus makes sense that `Microsoft.BCL.Async` would include `System.IO` and `System.Net`, to provide all the new `*Async()` methods.

If the question is different - why R# might suggest `System.Net` is needed...  @Lumirris I am not finding we can take R# as a reliable guide for what is needed.  For instance, on my machine it says `System.Runtime` is not needed for this csproj, but removing it causes compilation failure.  

I know 100% that Polly doesn't use `System.IO` and `System.Net` so I have removed from the proposed package.  I have tested and can run the most complicated [Polly-Samples](https://github.com/App-vNext/Polly-Samples) against the version with them removed.

**Comment by reisenberger** [comment]
>why System.Core is there, but not in the other Polly projects

`Microsoft.BCL.Async` helps a wide range of platforms adopt await/async (as described at https://www.nuget.org/packages/Microsoft.Bcl.Async/), including Silverlight.  In the reference sources for the LINQ extension methods, one sees code like:

```
 // Include Silverlight's managed resources
 #if SILVERLIGHT 
using System.Core; 
#endif
```

Perhaps more elements needed to be added to the package to support async in Silverlight, or Silverlight had some extension methods in slightly different namespaces? (_Hunching_; I haven't worked with Silverlight.  Happy for correction by anyone with detailed knowledge?)

I can remove `System.Core` and the NET40Async package compiles on my machine.  However, I have erred on the side of caution and left that one in the package, in case, for example, it is needed for Silverlight usage.

**Comment by reisenberger** [comment]
Hi @KBRETON1 Re hintpath and nuget.config:  https://docs.nuget.org/consume/nuget-config-settings confirms that $(Solutiondir)\Packages (which the hint paths reference) is the default _in absence of_ a `Nuget.Config` specifying a different value.

Do you possibly have the `repositoryPath` key overridden in your master/default `nuget.config` (located per [this](https://docs.nuget.org/consume/nuget-config-file))?  Could you confirm?  If so, could you confirm whether dropping a `Nuget.Config` as follows, next to `Polly.sln`, solves the problem for you? (with the original hintpaths in the csprojs _in place_).  Assuming this fixes for you, we should be able to add this to the repo.  Many thanks.

```xml
<configuration>
    <config>
        <add key="repositoryPath" value="./packages" />
    </config>
</configuration>
```

**Comment by KBRETON1** [comment]
@reisenberger, I searched my entire hard drive and only have <add key="repositorypath" value=".\Packages" />  entries in all of the nuget.config files.

Adding the nuget.config file to the src folder solved my issue.  I'm not sure why it went to the Appdata folder, but my issue is resolved with the work around.   No worries. 

Thanks for looking into it.

**Comment by KBRETON1** [comment]
@reisenberger, @Lumirris : This worked great on XP sp3 running .Net 40

**Comment by reisenberger** [comment]
@KBRETON1 Thanks, great to know!

**Comment by KBRETON1** [comment]
@reisenberger, what are the next steps and ultimately when this can be available in the nuget package?

Thanks,
Kevin

**Comment by reisenberger** [comment]
Hey @KBRETON1  I am on the road till Mon 6th, so only getting the odd hour at keyboard in hotel rooms this week.  Apologies for any delay and thanks for bearing with us on this! 

Remaining step is to adjust the build scripts to generate separate nuget packages for existing Polly and Polly.Net40Async.  I pushed an early version of build scripts yesterday to: https://github.com/reisenberger/Polly/tree/Net40Asyncv3 . @joelhulen on the App-vNext teams owns the nuget publishing gateway/process, so ...

@joelhulen I'll make a formal PR of this in a mo, and you can review / decide whether to go with the current build scripts to get the Polly.Net40Async packages out - or await refinement.  Thanks!

**Comment by joelhulen** [comment]
Thanks, @reisenberger. I Have been slammed busy, but I'll integrate your changes into the master build script.

**Comment by joelhulen** [comment]
@KBRETON1 @Lumirris @reisenberger 

I have combined @reisenberger's NuGet creation elements added to a separate cake file into the standard cake file that is executed by our AppVeyor build server. From that, we now have *four* NuGet packages: Polly.4.2.2.nupkg, Polly-Signed.4.2.2.nupkg, Polly.Net40Async.4.2.2.nupkg, and Polly.Net40Async-Signed.4.2.2.nupkg, respectively.

You can now [download the new Net40Async](https://www.nuget.org/packages/Polly.Net40Async/4.2.2) NuGet package, as well as its [signed counterpart](https://www.nuget.org/packages/Polly.Net40Async-Signed/4.2.2).

Please let me know if you have any issues with these NuGet packages.

**Comment by KBRETON1** [comment]
@joelhulen @Lumirris @reisenberger 

I believe there are some issues in the nuget.   Perhaps this comment should be part of #112

First, There are references added for Polly.dll and Polly.Net40Async.dll which causes this error: The type 'Policy' exists in both 'Polly.Net40Async, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' and 'Polly, Version=4.2.2.0, Culture=neutral, PublicKeyToken=null'

Secondly, when I manually remove Polly.dll in the csproj to avoid the duplication definition issue then the WaitAndRetryForeverAsync method is not found.  Using Object browser I can see the async methods are missing from Polly.Policy object.

Third, and this might not be an issue.  Shouldn't this new nuget package also install these nugets as dependencies?  Microsoft.BCL, Microsoft.Bcl.Build, Microsoft.Bcl.Async.

**Comment by Lumirris** [comment]
To your third point, Polly.Net40Async does take a dependency on
Microsoft.Bcl.Async, but I thought that Microsoft.Bcl.Async's dependencies
(Microsoft.Bcl.Build, and its dependency, Microsoft.BCL) would be
automatically pulled in, instead of having to list them explicitly.


Jeff

**Comment by joelhulen** [comment]
I believed this was the case as well. I'm sorry, I thought the package was already good to go. I'll see if I can track down why it's pulling in both DLLs, and why the Microsoft.Bcl.Async dependency isn't included.

**Comment by joelhulen** [comment]
@Lumirris @KBRETON1 

I have a new set of NuGet packages I'd like for you guys to test before I publish them to nuget.org. What's the best way for me to send them to you? Don't want to go through setting up myget or publishing them until I'm 100% certain they're ready.

Also, looking at the .nuspec file for the Polly.Net40Async NuGet (nupkg) file, it does list Microsoft.Bcl.Async as a dependent package. Not sure why the NuGet package manager didn't automatically pick this up for you.

**Comment by joelhulen** [comment]
[Polly 4.2.3 NuGet.zip](https://github.com/App-vNext/Polly/files/298364/Polly.4.2.3.NuGet.zip)

**Comment by joelhulen** [comment]
Try that. If you want, you can extract those packages to your local drive, then configure the NuGet Package Manager in VS to reference that folder as a NuGet package source. Let me know if you need help setting that up.

**Comment by Lumirris** [comment]
Ok, I'll take a look tonight! Thank you!


Jeff

**Comment by joelhulen** [comment]
@Lumirris Were you able to try it out? I want to update the NuGet packages ASAP. Thanks!

**Comment by KBRETON1** [comment]
@joelhulen @Lumirris
 
Thank you for your help!   :-)   We are making some progress. 

The first issue has been resolved.  The additional polly reference has been removed.   The other two issues still remain.  

2nd issue:  The async functions are missing from the Polly.Net40Async.dll.   I have opened the dll in Object Explorer and the new async methods are missing.   This might be a build/compile time issue as the async dll is missing the new methods.

3rd Issue: The bcl nugets are not installed as a dependency when install this nuget. In an unrelated BCL nuget package that my team created, we were able to set the BCL dependencies.   Installing our nuget forced the install of the other BCL nugets too.    Our nuget did not include any of the Microsoft.threading.*, System.Runtime.*, or System.Threading.* in the lib\net40 folder of the nuget package.   

I have a few screen shots but was not able to upload the png or zip files. :-(

**Comment by joelhulen** [comment]
@KBRETON1 

I really need to set up a dev environment to mimic yours...

As far as the BCL NuGets, you'll see the dependency listed in the `.nuspec` file for the Net40Async package as `<dependency id="Microsoft.Bcl.Async" version="1.0.168" />`. Is this not the correct version?

If you drag & drop the zip file to the comment box on this page, does it not upload the file?

**Comment by KBRETON1** [comment]
If I drag and drop a png or zip file it says it is not supported.  I'm trying this via email.     There is a screen shot of an unrelated nugget with the BCL dependencies that get installed.   I've also included a screen shot that shows the Async functions are missing from the dll.
Hope this works. :-) 

    On Saturday, June 4, 2016 5:21 PM, Joel Hulen <notifications@github.com> wrote:
 

 @KBRETON1 I really need to set up a dev environment to mimic yours...As far as the BCL NuGets, you'll see the dependency listed in the .nuspec file for the Net40Async package as <dependency id="Microsoft.Bcl.Async" version="1.0.168" />. Is this not the correct version?If you drag & drop the zip file to the comment box on this page, does it not upload the file?—

**Comment by joelhulen** [comment]
I see. You need to do this from the site in order for the drag/drop feature to work. No files came through...

**Comment by KBRETON1** [comment]
[bclscreenshots.zip](https://github.com/App-vNext/Polly/files/299077/bclscreenshots.zip)
Trying IE now.   
Screen shots show missing Async functions in object explorer.    
And an unrelated nuget package that has the BCL dependencies but doesn't include the actual dlls.

**Comment by joelhulen** [comment]
Hmmmm.... you're referencing this from within a PCL? I wonder if you need a PCL version of this library, which currently targets .NET 4.5, I believe.

**Comment by Lumirris** [comment]
Not yet. Probably not till Monday - busy this weekend. I'll let you know, though. 

Jeff

**Comment by joelhulen** [comment]
@KBRETON1 @Lumirris 

@reisenberger was gracious enough to apply a small fix to the production build that defines the SUPPORTSASYNC constant. Please try these new NuGet packages. Once verified, I'll publish them to nuget.org.

[Polly 4.2.3 NuGet packages.zip](https://github.com/App-vNext/Polly/files/303302/Polly.4.2.3.NuGet.packages.zip)

**Comment by KBRETON1** [comment]
@joelhulen @reisenberger @Lumirris 

I was able to compile and run with this nugget package.    :-)

It is still including the Microsoft.bcl assemblies directly in the lib folder.  I think it should have a dependency on those nuget packages instead of including the assemblies directly.   What are your thoughts?

**Comment by joelhulen** [comment]
@KBRETON1 @reisenberger @Lumirris 

Totally agree. @reisenberger Do you want to updated your project to exclude those from the build output and have it rely solely on the Microsoft.Bcl NuGet package to pull down those assemblies?

**Comment by reisenberger** [comment]
@joelhulen @KBRETON1 @Lumirris Also agree.  I believe in fact that three further layers of change were needed, now PRd in #108 : 

[1] **exclude non Polly dlls from the build output**: I've excluded them @joelhulen from the parts of the build output copied into the nuget package.  Is this what you meant, or have I misunderstood?
[2] **remove irrelevant dependencies**: See  PR, for details.
[3] **ensure stated dependencies targeted correct NET40 TFM/moniker**: again, see PR.

I tested this (compile-and-run) both against @KBRETON1 's ConsoleApplication2 , and an entirely fresh sample - ie without `Microsoft.BCL.Async` already included.  Calling in the Polly.Net40Async nuget package brought in the necessary dependencies, and allowed policies to be keyed, compiled and run.

@KBRETON1 Do you want to verify the package now both looks and behaves, as you would expect?

**Comment by joelhulen** [comment]
@KBRETON1 @Lumirris 

@reisenberger just cleaned up the dependencies. The BCL DLLs are no longer included and should now be automatically added via the NuGet dependency.

[Polly 4.2.3 NuGet packages take 5.zip](https://github.com/App-vNext/Polly/files/303589/Polly.4.2.3.NuGet.packages.take.5.zip)

**Comment by joelhulen** [comment]
@reisenberger Yes, perfect. The way you cleaned it up was exactly how I would have suggested. Thanks!

**Comment by KBRETON1** [comment]
@reisenberger @joelhulen @Lumirris 

This worked great!! Thank you.

There is a Polly.NET40Async.dll.config file in the package.  The other polly nuget's have polly.xml.  Not sure if this is needed or not.

**Comment by reisenberger** [comment]
@joelhulen @Lumirris @KBRETON1  I have done a line-by-line comparison of the `Polly.Net40.csproj` against `Polly.Net40Async.csproj`.  The moral of the story (for future similar cases...) could be that we will have a cleaner result/process, if we start (in similar circumstance) by copy-pasting `Polly.Net40.csproj` to `Polly.Net40Async.csproj`, rather than (say) adding a fresh project.

A shortly forthcoming PR:

+ re-instates XML file
+ ensures Polly.Async40.dll will be semver version-stamped (this had also been lost)
+ harmonizes other aspects

@KBRETON1 Thanks for the :eyes: .  Line-by-line comparison does not now reveal any other substantive differences - but let us know if you spot anything!

(comment on extra .config file coming separately)

**Comment by reisenberger** [comment]
@joelhulen / all.  The `Polly.NET40Async.dll.config` file is being added automatically by the build process [it doesn't stem from any app.config added manually to the solution], and it specifies binding redirects for two DLLs from `Microsoft.BCL.Async`.  

(This discussion)[http://stackoverflow.com/questions/16920370/] suggests it must be left in place, for a package based on BCL.Async to operate correctly.

**Comment by joelhulen** [comment]
Thanks for clarifying! Makes complete sense, as the dependency entries in the config contain the `bindingRedirect` properties with old and new assembly versions.

**Comment by joelhulen** [comment]
@KBRETON1 @Lumirris 

Attached is the latest (and hopefully last) round of NuGet packages with the latest cleanup and project tweaking results.

Please verify, and I'll post to nuget.org

[Polly 4.2.3 NuGet packages take 6.zip](https://github.com/App-vNext/Polly/files/303710/Polly.4.2.3.NuGet.packages.take.6.zip)

**Comment by KBRETON1** [comment]
@joelhulen @reisenberger @Lumirris 

This version worked great.  It built fine and ran successfully on my XP with .Net 4.0 image.   

I need to do some research about Binding Redirects as I my experience has been in the app.config only and not at the assembly level.  Is the Polly.Net40Async.config supposed to be redistributed or is this how the bindredirects get created in the app.config?   The ConsoleApplication2 running on XP just has the app.config with the binding redirects.

Thank you.

**Comment by KBRETON1** [comment]
I left out that I believe the app.config is modified with the redirects from the dependency on the microsoft.bcl.* nugets.  This might mean that the Polly.Net40Async.config shouldn't be included in the nuget.

**Comment by joelhulen** [comment]
@KBRETON1 I've discussed this offline with @reisenberger, but I want to know whether you agree, since you're closer to this particular issue.

We've tested adding the NuGet package to a simple app, both with and without the config file. It worked in both cases. The [discussion](http://stackoverflow.com/questions/16920370/) @reisenberger linked to earlier, which suggest that the config file must be kept in place for BCL.Async to function correctly, makes me more confident about leaving it in, which doesn't appear to hurt anything, rather than taking it out with unknown consequences (such as dynamically loading assemblies, or other scenarios we haven't tested).

What I've seen in the past in similar situations, is where the included config file is used if the containing project does not have one. Otherwise, the containing project's config is used.

I'm not sure if it works exactly the same way in this scenario, but it seems as though the config file from our NuGet package would inform the containing project as to what the dependency values should be.

In the meantime, I've gone ahead and published these new (4.2.3) packages to nuget.org.
